File: pippip-native-ios/Support/AccountManager.m
//
//  AccountManager.m
//  pippip-native-ios
//
//  Created by Steve Brenneis on 10/9/17.
//  Copyright © 2017 seComm. All rights reserved.
//

#import "AccountManager.h"
#import "AccountConfig.h"
#import <Realm/Realm.h>

static const float CURRENT_VERSION = 1.0;
static NSString *acctName = nil;

@interface AccountManager ()
{
}

@end

@implementation AccountManager

+ (NSString*)accountName {
    return acctName;
}

+ (void)accountName:(NSString*)name {
    acctName = name;
}

+ (BOOL)production {
    return true;
}

- (void)loadAccount {

    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *docPath = [paths objectAtIndex:0];
    NSString *vaultsPath = [docPath stringByAppendingPathComponent:@"PippipVaults"];
    NSArray *vaultNames = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:vaultsPath
                                                                              error:nil];
    NSMutableArray *accountNames = [vaultNames mutableCopy];
#if TARGET_OS_SIMULATOR
    if (accountNames.count > 0) {
        [accountNames removeObject:@".DS_Store"];
    }
#endif
    if (accountNames.count > 0) {
        acctName = [accountNames firstObject];
        [self loadConfig];
    }

}

- (void)loadConfig {

    [self setRealmConfiguration];
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"accountName = %@", acctName];
    RLMResults<AccountConfig*> *configList = [AccountConfig objectsWithPredicate:predicate];
    if (configList.count > 0) {
        // This is where we do version migrations
        AccountConfig *config = [configList firstObject];
        if (config.version < 1.0) {
            RLMRealm *realm = [RLMRealm defaultRealm];
            [realm beginWriteTransaction];
            config.cleartextMessages = NO;
            [realm commitWriteTransaction];
        }
        if (config.version < 1.1) {
            RLMRealm *realm = [RLMRealm defaultRealm];
            [realm beginWriteTransaction];
            config.localAuth = YES;
            config.version = CURRENT_VERSION;
            [realm commitWriteTransaction];
        }
    }

}

- (void)setDefaultConfig {

    [self setRealmConfiguration];
    AccountConfig *config = [[AccountConfig alloc] init];
    config.version = 1.0;
    config.accountName = acctName;
    config.contactPolicy = @"whitelist";
    config.messageId = 1;
    config.contactId = 1;
    config.whitelist = nil;
    config.idMap = nil;
    config.cleartextMessages = NO;

    RLMRealm *realm = [RLMRealm defaultRealm];
    [realm beginWriteTransaction];
    [realm addObject:config];
    [realm commitWriteTransaction];

}

- (void)setRealmConfiguration {
    
    RLMRealmConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
    // Use the default directory, but replace the filename with the username
    config.fileURL = [[[config.fileURL URLByDeletingLastPathComponent]
                       URLByAppendingPathComponent:acctName]
                      URLByAppendingPathExtension:@"realm"];
    config.schemaVersion = 9;
    config.migrationBlock = ^(RLMMigration *migration, uint64_t oldSchemaVersion) {
        if (oldSchemaVersion < 9) {
            // No migration necessary
        }
    };
    [RLMRealmConfiguration setDefaultConfiguration:config];
    
}

@end

File: pippip-native-ios/Support/AccountManager.swift
//
//  AccountManager.swift
//  pippip-native-ios
//
//  Created by Steve Brenneis on 6/10/18.
//  Copyright © 2018 seComm. All rights reserved.
//

import UIKit
import RealmSwift

class AccountManager: NSObject {

    static let production = true
    static var accountName: String?

    func loadAccount() throws {
        
        let fileManager = FileManager.default
        let documentDirectory = try fileManager.url(for: .documentDirectory,
                                                    in: .userDomainMask,
                                                    appropriateFor:nil,
                                                    create:false)
        let vaultsURL = documentDirectory.appendingPathComponent("PippipVaults")
        let path = vaultsURL.path as NSString
        if !fileManager.fileExists(atPath: path.expandingTildeInPath) {
            try fileManager.createDirectory(at: vaultsURL, withIntermediateDirectories: false, attributes: nil)
        }
        let vaults = try fileManager.contentsOfDirectory(at: vaultsURL,
                                                         includingPropertiesForKeys: nil,
                                                         options: .skipsHiddenFiles)
        if !vaults.isEmpty {
            AccountManager.accountName = vaults[0].lastPathComponent
        }
        
    }

    func loadConfig() {
        
        setRealmConfiguration()
        let realm = try! Realm()
        let config = realm.objects(AccountConfig.self).first
        print("AccountConfig version \(config!.version)")
        
    }

    func setDefaultConfig() {
        
        let config = AccountConfig()
        do {
            let realm = try Realm()
            try realm.write {
                realm.add(config)
            }
        }
        catch {
            print("Error writing initial config to database: \(error)")
        }
        
    }

    func setRealmConfiguration() {
        
        var realmConfig = Realm.Configuration()
        realmConfig.fileURL = realmConfig.fileURL?.deletingLastPathComponent()
            .appendingPathComponent("\(AccountManager.accountName!).realm")
        realmConfig.schemaVersion = 13
        realmConfig.migrationBlock = { (migration, oldSchemaVersion) in
            // Schema version 13 is Realm Swift
            if oldSchemaVersion < 13 {
                migration.enumerateObjects(ofType: DatabaseContact.className()) { (oldObject, newObject) in
                    newObject!["contactId"] = oldObject!["contactId"]
                    newObject!["encoded"] = oldObject!["encoded"]
                }
                migration.enumerateObjects(ofType: DatabaseContactRequest.className()) { (oldObject, newObject) in
                    newObject!["publicId"] = oldObject!["publicId"]
                    newObject!["nickname"] = oldObject!["nickname"]
                }
                migration.enumerateObjects(ofType: DatabaseMessage.className()) { (oldObject, newObject) in
                    newObject!["version"] = oldObject!["version"]
                    newObject!["contactId"] = oldObject!["contactId"]
                    newObject!["messageId"] = oldObject!["messageId"]
                    newObject!["ciphertext"] = oldObject!["message"]
                    newObject!["keyIndex"] = oldObject!["keyIndex"]
                    newObject!["sequence"] = oldObject!["sequence"]
                    newObject!["timestamp"] = oldObject!["timestamp"]
                    newObject!["cleartext"] = oldObject!["cleartext"]
                    newObject!["read"] = oldObject!["read"]
                    newObject!["acknowledged"] = oldObject!["acknowledged"]
                    newObject!["originating"] = oldObject!["sent"]
                    newObject!["compressed"] = oldObject!["compressed"]
                    newObject!["failed"] = oldObject!["failed"]
                }
                migration.enumerateObjects(ofType: AccountConfig.className()) { (oldObject, newObject) in
                    newObject!["version"] = oldObject!["version"]
                    newObject!["nickname"] = oldObject!["nickname"]
                    newObject!["contactPolicy"] = oldObject!["contactPolicy"]
                    newObject!["currentMessageId"] = oldObject!["messageId"]
                    newObject!["currentContactId"] = oldObject!["contactId"]
                    newObject!["whitelist"] = oldObject!["whitelist"]
                    newObject!["storeCleartextMessages"] = oldObject!["cleartextMessages"]
                    newObject!["useLocalAuth"] = oldObject!["localAuth"]
                }
            }
        }
        Realm.Configuration.defaultConfiguration = realmConfig
        
    }

}
